### PR TITLE
WIP, NEP: add roadmap process NEP

### PR DESCRIPTION
**Pull request by mattip**
Still a rough draft, needs more community input and internal organization. I would like to get it into the NEP process so we can keep track of discussions and comments

### CODE REVIEW

File: doc/neps/nep-0022-roadmap.rst [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+
+Philosophy / scope
+------------------
+
+What is part of NumPy, what is not?
+
+- In memory, N-dimensional single pointer + strided arrays on CPUs
+
+  - Not specialized hardware
+- Basic functionality for scientific computing
```
Comment (eric-wieser) [inline doc/neps/nep-0022-roadmap.rst]: This rst syntax is a little weird - can't you just write

```
- outer list
  - inner list
- outer list
```
or maybe
```
- outer list
  - inner list

- outer list
```

**Comment by rgommers** [comment]
I don't think the roadmap should be a NEP. NEPs are about individual ideas/proposals, and are frozen at some point after accept/reject. The roadmap should be an evolving plan/document. 

Agree that it's good to have it version controlled so we can track things, but preferably somewhere else (either a separate section in the docs, a separate repo, or part of the numpy.org repo).

File: doc/neps/nep-0022-roadmap.rst (hunk 1 above)
Comment (mattip) [reply doc/neps/nep-0022-roadmap.rst]: RST must have a blank line before the next identation level, so I can do the last one

**Comment by stefanv** [comment]
Hi @rgommers;

Our thinking is that this is a meta-NEP of sorts, and thus falls into the Process category, as outlined in [NEP0](http://www.numpy.org/neps/nep-0000.html).  NEP0 states that "Process NEPs may be updated over time to reflect changes to development practices and other details. The precise process followed in these cases will depend on the nature and purpose of the NEP being updated.".

We've been using NEPs more heavily recently, so these waters are still being tested, and while it can be argued that a NEP is a good fit for a roadmap, we don't have any specific interest in making it so.  Alternative suggestions are welcome!

**Comment by rgommers** [comment]
Good point Stefan. A process NEP does sound useful. I'd think that that should be along the lines of:
- We have a NumPy Roadmap. It contains items of type X, Y, and Z.
- The process for updating the roadmap is <etc.>
But it's a bit odd to include the actual items on the roadmap themselves, since those are changing over time.

**Comment by stefanv** [comment]
Another option Matti suggested was that we create a new repo with NumPy organization material, where the merge threshold is much lower.  I'm not sure the roadmap will have enough visibility in such a place, but all the options on the table seem like a bit of a strange fit anyway.

**Comment by rgommers** [comment]
That sounds fine to me as well. We can link the roadmap from the docs and from the numpy.org front page, so visibility should be okay then I think.

**Comment by mhvk** [comment]
A wiki page with appropriate links perhaps?

**Comment by stefanv** [comment]
We currently have it as a wiki page, but while the history is available, it seems harder to track, whereas when it is in a code repository there are explicit PRs.  Wiki pages also are harder to discover (since we don't have any other documentation there), and tend to become outdated.

File: doc/neps/nep-0022-roadmap.rst [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+
+    - linear algebra
+    - FFT
+    - Windowing functions for ffts.
+    - where is the cutoff? Simple things that everyone needs should be in NumPy, specialized and complicated functions in SciPy. How do we decide which is which?
+  - Infrastructrure for data science packages (distutils, f2py, testing)
+- Higher level APIs for N-dimensional arrays
+
+  - NumPy is a *de facto* standard for array APIs in Python
+  - Protocols like `__array_ufunc__`, `__array_func__`.
```
Comment (shoyer) [inline doc/neps/nep-0022-roadmap.rst]: `__array_func__` -> `__array_function__`?

File: doc/neps/nep-0022-roadmap.rst [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+  - Challenge: retention; if we spend time training new contributors, how do we get them to stay 5 years instead of 1 or 2 (grad school, e.g.)
+  - Engage with universities or industry (who could e.g. sponsor developer time)
+
+    - Can we have a standard mechanisms whereby industry can engage
+
+      - Sponsorship of: developer time / specific feature / ... ?
+      - Could take on self-contained tasks, such as type stubs
+      - Companies may benefit from smaller, better scoped packages; they may not
+        want to take on "NumPy", but could be willing to engage with a smaller project
+        that won't become a time sink. Hypothetical example: company like UK Met
```
Comment (shoyer) [inline doc/neps/nep-0022-roadmap.rst]: nit: UK Met Office is an (governmental?) organization, not a company

**Comment by shoyer** [comment]
My slight preference would be to keep this as a NEP (even if we update it continuously) rather than somewhere else. In my mind, NEP really serve as more of a centralized repository of design documents, rather than a more limited list of "enhancement proposals."

**Comment by mhvk** [comment]
Since it is submitted as a NEP, might as well do that first.  Wonder slightly about the numbering, though - should they have a separate track?

**Review by rgommers — CHANGES_REQUESTED** [review]

**Comment by rgommers** [comment]
@shoyer and I caught up about this yesterday. We were thinking about slightly different things - Stephan wants to ensure that the roadmap ends up on http://www.numpy.org/neps/; it doesn't need to be a NEP.

We also spent some time drafting a v2 of the roadmap, based on the extended version of this PR from the BoF and some more thoughts of our own. Plus a scope refactoring - there's a difference between a roadmap and "things we want/need to work on".

More to follow soon.

**Comment by shoyer** [comment]
See https://github.com/numpy/numpy/wiki/NumPy-roadmap-v2 for the raw notes from @rgommers  and myself (not very different from this yet!).

**Comment by rgommers** [comment]
Review comments were not processed and are now superceded by the edits on the v2 roadmap as well as the BoF brainstorm output. So I propose to close this PR and go back to wiki edits.

**Comment by mattip** [comment]
We should set a time frame for completing the rewrite

**Comment by rgommers** [comment]
> We should set a time frame for completing the rewrite

good idea. how about a complete draft ready for review by the end of next week? plus by then also a plan/proposal for the format and for getting the reviewed version integrated in the docs (or numpy.org)?

**Comment by stefanv** [comment]
Instead of closing, how about making a commit here with the latest version,
and keep refining it?

**Comment by rgommers** [comment]
collaborative editing on a PR is very cumbersome, so can we instead reopen it once we have something that is ready for review next week?

**Comment by stefanv** [comment]
Sure, that should work just fine.  In the mean time, I will work on a home for it in the neps docs.

**Pull request closed without merging**